In [2]:
import numpy as np 
import pandas as pd
import seaborn as sns
from Logistic_Regression import LogisticRegression
from warnings import filterwarnings

filterwarnings('ignore')

# Loading titanic dataset from seaborn

In [3]:
data = sns.load_dataset('titanic')

# EDA, Cleaning and preproccessing 


In [4]:
data.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [5]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB


In [6]:
data = data.drop(['deck', 'embark_town', 'who','alive'], axis = 1)

In [7]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   survived    891 non-null    int64   
 1   pclass      891 non-null    int64   
 2   sex         891 non-null    str     
 3   age         714 non-null    float64 
 4   sibsp       891 non-null    int64   
 5   parch       891 non-null    int64   
 6   fare        891 non-null    float64 
 7   embarked    889 non-null    str     
 8   class       891 non-null    category
 9   adult_male  891 non-null    bool    
 10  alone       891 non-null    bool    
dtypes: bool(2), category(1), float64(2), int64(4), str(2)
memory usage: 58.6 KB


In [8]:
data['age'] = data['age'].fillna(data['age'].mean(), inplace = True)

In [9]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   survived    891 non-null    int64   
 1   pclass      891 non-null    int64   
 2   sex         891 non-null    str     
 3   age         891 non-null    float64 
 4   sibsp       891 non-null    int64   
 5   parch       891 non-null    int64   
 6   fare        891 non-null    float64 
 7   embarked    889 non-null    str     
 8   class       891 non-null    category
 9   adult_male  891 non-null    bool    
 10  alone       891 non-null    bool    
dtypes: bool(2), category(1), float64(2), int64(4), str(2)
memory usage: 58.6 KB


# Encoding

In [10]:
data['sex'] = data['sex'].map({
    'male':1,
    'female': 0
}, inplace = True)

data = data.rename(columns={'sex':'is_male'})

In [11]:
data

,survived,pclass,is_male,age,sibsp,parch,fare,embarked,class,adult_male,alone
0,0,3,1,22.000000,1,0,7.2500,S,Third,True,False
1,1,1,0,38.000000,1,0,71.2833,C,First,False,False
2,1,3,0,26.000000,0,0,7.9250,S,Third,False,True
3,1,1,0,35.000000,1,0,53.1000,S,First,False,False
4,0,3,1,35.000000,0,0,8.0500,S,Third,True,True
...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,1,27.000000,0,0,13.0000,S,Second,True,True
887,1,1,0,19.000000,0,0,30.0000,S,First,False,True
888,0,3,0,29.699118,1,2,23.4500,S,Third,False,False
889,1,1,1,26.000000,0,0,30.0000,C,First,True,True


In [12]:
data = pd.get_dummies(data, columns=['embarked','class'], dtype= int)

In [13]:
data.head()

,survived,pclass,is_male,age,sibsp,parch,fare,adult_male,alone,embarked_C,embarked_Q,embarked_S,class_First,class_Second,class_Third
0,0,3,1,22.0,1,0,7.2500,True,False,0,0,1,0,0,1
1,1,1,0,38.0,1,0,71.2833,False,False,1,0,0,1,0,0
2,1,3,0,26.0,0,0,7.9250,False,True,0,0,1,0,0,1
3,1,1,0,35.0,1,0,53.1000,False,False,0,0,1,1,0,0
4,0,3,1,35.0,0,0,8.0500,True,True,0,0,1,0,0,1


In [14]:
data[['adult_male','alone']] = data[['adult_male','alone']].astype(int)

In [15]:
data.head()

,survived,pclass,is_male,age,sibsp,parch,fare,adult_male,alone,embarked_C,embarked_Q,embarked_S,class_First,class_Second,class_Third
0,0,3,1,22.0,1,0,7.2500,1,0,0,0,1,0,0,1
1,1,1,0,38.0,1,0,71.2833,0,0,1,0,0,1,0,0
2,1,3,0,26.0,0,0,7.9250,0,1,0,0,1,0,0,1
3,1,1,0,35.0,1,0,53.1000,0,0,0,0,1,1,0,0
4,0,3,1,35.0,0,0,8.0500,1,1,0,0,1,0,0,1


In [16]:
X = data.drop('survived', axis = 1)
y = data['survived']

# Test train split

In [17]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)  

# Scalling

In [18]:
from sklearn.preprocessing import StandardScaler

scalar = StandardScaler()

cols = ['pclass','age','parch','fare','sibsp']

X_train[cols] = scalar.fit_transform(X_train[cols])

X_test[cols] = scalar.transform(X_test[cols])

In [19]:
X_train.head()

,pclass,is_male,age,sibsp,parch,fare,adult_male,alone,embarked_C,embarked_Q,embarked_S,class_First,class_Second,class_Third
331,-1.614136,1,1.229207,-0.470722,-0.479342,-0.078684,1,1,0,0,1,1,0,0
733,-0.400551,1,-0.503505,-0.470722,-0.479342,-0.377145,1,1,0,0,1,0,1,0
382,0.813034,1,0.189580,-0.470722,-0.479342,-0.474867,1,1,0,0,1,0,0,1
704,0.813034,1,-0.272477,0.379923,-0.479342,-0.476230,1,0,0,0,1,0,0,1
813,0.813034,0,-1.812666,2.931860,2.048742,-0.025249,0,0,0,0,1,0,0,1


In [20]:
X_test.head()

,pclass,is_male,age,sibsp,parch,fare,adult_male,alone,embarked_C,embarked_Q,embarked_S,class_First,class_Second,class_Third
709,0.813034,1,0.012390,0.379923,0.784700,-0.333901,1,0,1,0,0,0,0,1
439,-0.400551,1,0.112570,-0.470722,-0.479342,-0.425284,1,1,0,0,1,0,1,0
840,0.813034,1,-0.734533,-0.470722,-0.479342,-0.474867,1,1,0,0,1,0,0,1
720,-0.400551,0,-1.812666,-0.470722,0.784700,0.007966,0,0,0,0,1,0,1,0
39,0.813034,0,-1.196590,0.379923,-0.479342,-0.411002,0,0,1,0,0,0,0,1


# Custom Model 

In [21]:
model = LogisticRegression()

model.fit(X_train, y_train)

prediction = model.predict(X_test)

# Model Evaluation

In [22]:
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, confusion_matrix

In [23]:
confusion_matrix(prediction, y_test)

array([[91, 20],
       [14, 54]])

In [38]:
accuracy = accuracy_score(prediction, y_test)
precision = precision_score(prediction, y_test)
recall = recall_score(prediction, y_test)
f1 = f1_score(prediction, y_test)

In [39]:
print(f'''accuracy: {accuracy*100} %
      
precision: {precision*100} %

recall: {recall*100} %

f1: {f1*100} %''')

accuracy: 81.00558659217877 %

precision: 72.97297297297297 %

recall: 79.41176470588235 %

f1: 76.05633802816901 %


# Comparison with sklearn model

In [26]:
from sklearn.linear_model import LogisticRegression

In [27]:
sk_model = LogisticRegression()

sk_model.fit(X_train,y_train)

sk_prediction = sk_model.predict(X_test)

In [28]:
sk_prediction

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 1, 1])

In [30]:
confusion_matrix(sk_prediction, y_test)

array([[91, 19],
       [14, 55]])

In [40]:
sk_accuracy = accuracy_score(sk_prediction, y_test)
sk_precision = precision_score(sk_prediction, y_test)
sk_recall = recall_score(sk_prediction, y_test)
sk_f1 = f1_score(sk_prediction, y_test)

In [41]:
print(f'''sk_accuracy: {sk_accuracy*100} %
      
sk_precision: {sk_precision*100} %

sk_recall: {sk_recall*100} %

sk_f1: {sk_f1*100} %''')

sk_accuracy: 81.56424581005587 %

sk_precision: 74.32432432432432 %

sk_recall: 79.71014492753623 %

sk_f1: 76.92307692307693 %


# Comparison DataFrame 

In [42]:
comparison_df = pd.DataFrame(
    {
        'accuracy' : [accuracy,sk_accuracy],
        'precision' : [precision, sk_precision],
        'recall' : [recall, sk_recall],
        'f1_score' : [f1,sk_f1]
    },
    index= ['custom_model','sk_model']
)

In [43]:
comparison_df

,accuracy,precision,recall,f1_score
custom_model,0.810056,0.729730,0.794118,0.760563
sk_model,0.815642,0.743243,0.797101,0.769231


# Conclusion

In this project, I implemented **Logistic Regression from scratch using NumPy**, without relying on scikit-learn's `LogisticRegression` class. The implementation helped me understand the core concepts of Logistic Regression, including the **Sigmoid Function, Log Loss, Gradient Descent, parameter updates, and binary classification**.

After preprocessing the Titanic dataset, the custom model achieved **81.01% Accuracy**, **72.97% Precision**, **79.41% Recall**, and **76.06% F1 Score**.

To validate the implementation, the custom model was compared with **scikit-learn's LogisticRegression** model:

| Metric        | Custom Model | Scikit-learn Model |
| ------------- | -----------: | -----------------: |
| **Accuracy**  |   **81.01%** |         **81.56%** |
| **Precision** |   **72.97%** |         **74.32%** |
| **Recall**    |   **79.41%** |         **79.71%** |
| **F1 Score**  |   **76.06%** |         **76.92%** |

The results are very close to the scikit-learn implementation, confirming that the scratch model was implemented correctly. Although the scikit-learn model achieved slightly better performance, the custom implementation provided valuable hands-on experience in building and understanding Logistic Regression from first principles.

